File: server/excelService.ts
/**
 * ExcelService — Ported from com.icare.service.ExcelService (Java)
 *
 * Reads Hospital_DB.xlsx, parses the Reference Table for hospitals and
 * individual doctor sheets, links doctors to hospitals via reference key,
 * and caches everything in memory.
 */

import { createRequire } from "module";
import path from "path";
import fs from "fs";
import { fileURLToPath } from "url";

const require = createRequire(import.meta.url);
const XLSX = require("xlsx");

// ─── Interfaces ──────────────────────────────────────────────────

export interface Doctor {
  id: number;
  name: string;
  type: string;
  department: string;
  schedule: string;
  priceRange: string;
  hmos: string[];
  contact: string;
  secretary: string;
  tags: string[];
}

export interface Hospital {
  id: string;
  numericId: number;
  name: string;
  address: string;
  latitude: number;
  longitude: number;
  services: string[];
  contactNumber: string;
  operatingHours: string;
  philHealthStatus: string;
  referenceKey: string;
  level: string;
  doctors: Doctor[];
}

// ─── Constants ───────────────────────────────────────────────────

const REFERENCE_SHEET_NAME = "Reference Table";

// ─── Cache ───────────────────────────────────────────────────────

let hospitalCache: Hospital[] | null = null;

// ─── Header Normalization ────────────────────────────────────────

function normalizeHeader(raw: string): string {
  return (raw || "").toLowerCase().replace(/[\s_\-]/g, "");
}

function normalizeTag(value: string): string {
  return value.trim().toLowerCase().replace(/[^a-z0-9]+/g, "-");
}

// ─── Safe Cell Readers ───────────────────────────────────────────

function safeString(cell: any): string {
  if (!cell) return "";
  if (cell.t === "s") return String(cell.v ?? "").trim();
  if (cell.t === "n") return String(cell.v ?? "").trim();
  if (cell.t === "b") return String(cell.v ?? "").trim();
  return "";
}

function safeDouble(cell: any): number {
  if (!cell) return 0.0;
  if (cell.t === "n") return Number(cell.v) || 0.0;
  const text = safeString(cell);
  if (!text) return 0.0;
  const parsed = parseFloat(text);
  return isNaN(parsed) ? 0.0 : parsed;
}

// ─── Column Index Lookup ─────────────────────────────────────────

function getColumnIndex(headers: string[], fieldName: string, fallback: number): number {
  const idx = headers.indexOf(fieldName);
  return idx >= 0 ? idx : fallback;
}

function buildHeaderNames(sheet: any, row: number): string[] {
  const headers: string[] = [];
  const range = XLSX.utils.decode_range(sheet["!ref"] || "A1");
  for (let col = range.s.c; col <= range.e.c; col++) {
    const addr = XLSX.utils.encode_cell({ r: row, c: col });
    headers.push(normalizeHeader(safeString(sheet[addr])));
  }
  return headers;
}

function getCellAt(sheet: any, row: number, col: number): any {
  const addr = XLSX.utils.encode_cell({ r: row, c: col });
  return sheet[addr];
}

// ─── HMO Parsing ─────────────────────────────────────────────────

function parseHmos(cell: any): string[] {
  const raw = safeString(cell);
  if (!raw) return [];
  return raw.split(",").map(s => s.trim()).filter(s => s.length > 0);
}

// ─── Doctor Tag Builder ──────────────────────────────────────────

function buildDoctorTags(department: string, specializations: string, type: string): string[] {
  const tags: string[] = [];

  if (department && department.trim()) {
    department.split(",").map(s => s.trim()).filter(s => s).forEach(s => tags.push(normalizeTag(s)));
  }
  if (specializations && specializations.trim()) {
    specializations.split(",").map(s => s.trim()).filter(s => s).forEach(s => tags.push(normalizeTag(s)));
  }
  if (type && type.trim()) {
    tags.push(normalizeTag(type));
  }
  if (tags.length === 0) {
    tags.push("general");
  }

  return [...new Set(tags)];
}

// ─── Doctor ID Generator ─────────────────────────────────────────

function generateDoctorId(sheetName: string, name: string, department: string): number {
  // Simple hash to replicate Java's Objects.hash()
  let hash = 0;
  const str = `${sheetName}:${name}:${department}`;
  for (let i = 0; i < str.length; i++) {
    hash = ((hash << 5) - hash + str.charCodeAt(i)) | 0;
  }
  return Math.abs(hash);
}

// ─── Hospital Services Builder ───────────────────────────────────

function buildHospitalServices(doctors: Doctor[], hospitalType: string): string[] {
  if (!doctors || doctors.length === 0) {
    return [hospitalType && hospitalType.trim() ? hospitalType : "General"];
  }
  const departments = doctors
    .map(d => d.department)
    .filter(d => d && d.trim())
    .map(d => d.trim());
  return [...new Set(departments)];
}

// ─── Sheet Parsers ───────────────────────────────────────────────

function isDoctorSheet(sheet: any): boolean {
  const headers = buildHeaderNames(sheet, 0);
  const first = headers[0] || "";
  return first.includes("nameofdoctor") || first.includes("doctor");
}

function parseDoctorSheet(sheet: any, sheetName: string): Doctor[] {
  const headers = buildHeaderNames(sheet, 0);
  const range = XLSX.utils.decode_range(sheet["!ref"] || "A1");
  const doctors: Doctor[] = [];

  for (let row = 1; row <= range.e.r; row++) {
    const nameCol = getColumnIndex(headers, "nameofdoctor", 0);
    const name = safeString(getCellAt(sheet, row, nameCol));
    if (!name) continue;

    const typeCol = getColumnIndex(headers, "typeofdoctor", 1);
    const deptCol = getColumnIndex(headers, "department", 2);
    const otherSpecCol = getColumnIndex(headers, "otherspecializations", 3);
    const schedCol = getColumnIndex(headers, "schedule", 4);
    const priceCol = getColumnIndex(headers, "price", 7);
    const hmoCol = getColumnIndex(headers, "hmos", 6);
    const contactCol = getColumnIndex(headers, "contactnumber", 9);
    const secCol = getColumnIndex(headers, "medicalsecretary", 8);

    const department = safeString(getCellAt(sheet, row, deptCol));
    const otherSpecializations = safeString(getCellAt(sheet, row, otherSpecCol));
    let type = safeString(getCellAt(sheet, row, typeCol));
    if (!type) type = "General Practitioner";

    const doctor: Doctor = {
      id: generateDoctorId(sheetName, name, department),
      name,
      type,
      department,
      schedule: safeString(getCellAt(sheet, row, schedCol)),
      priceRange: safeString(getCellAt(sheet, row, priceCol)),
      hmos: parseHmos(getCellAt(sheet, row, hmoCol)),
      contact: safeString(getCellAt(sheet, row, contactCol)),
      secretary: safeString(getCellAt(sheet, row, secCol)),
      tags: buildDoctorTags(department, otherSpecializations, type),
    };

    doctors.push(doctor);
  }

  return doctors;
}

function isReferenceRowEmpty(sheet: any, row: number, maxCol: number): boolean {
  for (let col = 0; col <= maxCol; col++) {
    if (safeString(getCellAt(sheet, row, col)).trim()) return false;
  }
  return true;
}

// ─── Main Loading Logic ──────────────────────────────────────────

function resolveXlsxPath(): string {
  const isVercel = process.env.VERCEL === "1";
  console.log(`[ExcelService] Detecting environment... (Vercel: ${isVercel})`);
  
  // Try multiple locations for flexibility
  const currentDir = path.dirname(fileURLToPath(import.meta.url));
  const candidates = [
    path.resolve(process.cwd(), "data", "Hospital_DB.xlsx"),
    path.resolve(process.cwd(), "Hospital_DB.xlsx"), // fallback root
    path.resolve(currentDir, "..", "data", "Hospital_DB.xlsx"),
  ];

  for (const candidate of candidates) {
    if (fs.existsSync(candidate)) {
      return candidate;
    }
  }

  throw new Error(
    `Hospital_DB.xlsx not found. Searched:\n${candidates.join("\n")}`
  );
}

function loadHospitals(): Hospital[] {
  const filePath = resolveXlsxPath();
  console.log(`[ExcelService] Loading hospital data from: ${filePath}`);

  const workbook = XLSX.readFile(filePath);
  const sheetMap = new Map<string, any>();

  for (const name of workbook.SheetNames) {
    sheetMap.set(name.trim(), workbook.Sheets[name]);
  }

  // Find reference sheet
  let referenceSheet: any = null;
  let referenceSheetName = "";

  if (sheetMap.has(REFERENCE_SHEET_NAME)) {
    referenceSheet = sheetMap.get(REFERENCE_SHEET_NAME)!;
    referenceSheetName = REFERENCE_SHEET_NAME;
  } else {
    // Fallback: find sheet with healthcare facility header
    for (const [name, sheet] of sheetMap.entries()) {
      const headers = buildHeaderNames(sheet, 0);
      if (headers[0]?.includes("nameofhealthcarefacility") || headers[0]?.includes("healthcarefacility")) {
        referenceSheet = sheet;
        referenceSheetName = name;
        break;
      }
    }
  }

  if (!referenceSheet) {
    console.warn("[ExcelService] No reference sheet found in workbook");
    return [];
  }

  // Parse doctor sheets
  const doctorsByReference = new Map<string, Doctor[]>();
  for (const [name, sheet] of sheetMap.entries()) {
    if (name === referenceSheetName) continue;
    if (isDoctorSheet(sheet)) {
      const doctors = parseDoctorSheet(sheet, name);
      if (doctors.length > 0) {
        doctorsByReference.set(name, doctors);
      }
    }
  }

  // Parse reference table
  const headers = buildHeaderNames(referenceSheet, 0);
  const range = XLSX.utils.decode_range(referenceSheet["!ref"] || "A1");
  const hospitals: Hospital[] = [];

  for (let row = 1; row <= range.e.r; row++) {
    if (isReferenceRowEmpty(referenceSheet, row, range.e.c)) continue;

    const refKeyCol = getColumnIndex(headers, "referencekey", 5);
    const referenceKey = safeString(getCellAt(referenceSheet, row, refKeyCol));

    let name = safeString(getCellAt(referenceSheet, row, getColumnIndex(headers, "nameofhealthcarefacility", 0)));
    if (!name) {
      name = safeString(getCellAt(referenceSheet, row, getColumnIndex(headers, "name", 0)));
    }
    if (!name) continue;

    const type = safeString(getCellAt(referenceSheet, row, getColumnIndex(headers, "type", 1)));
    const level = safeString(getCellAt(referenceSheet, row, getColumnIndex(headers, "level", 2)));
    const latitude = safeDouble(getCellAt(referenceSheet, row, getColumnIndex(headers, "latitude", 3)));
    const longitude = safeDouble(getCellAt(referenceSheet, row, getColumnIndex(headers, "longitude", 4)));
    const address = safeString(getCellAt(referenceSheet, row, getColumnIndex(headers, "address", 2)));

    // Skip rows with no valid coordinates
    if (latitude === 0.0 && longitude === 0.0) continue;

    const doctors = doctorsByReference.get(referenceKey) || [];
    const services = buildHospitalServices(doctors, type);

    const hospital: Hospital = {
      id: referenceKey || String(row),
      numericId: row,
      name,
      address,
      latitude,
      longitude,
      services,
      contactNumber: "",
      operatingHours: "",
      philHealthStatus: "",
      referenceKey,
      level,
      doctors,
    };

    hospitals.push(hospital);
  }

  console.log(`[ExcelService] Loaded ${hospitals.length} hospitals, ${doctorsByReference.size} doctor sheets`);
  return hospitals;
}

// ─── Public API ──────────────────────────────────────────────────

export function initExcelService(): void {
  hospitalCache = loadHospitals();
}

export function getHospitals(): Hospital[] {
  if (!hospitalCache) {
    hospitalCache = loadHospitals();
  }
  return [...hospitalCache];
}

export function findHospitalById(hospitalId: string): Hospital | undefined {
  return getHospitals().find(
    h => h.id === hospitalId || String(h.numericId) === hospitalId
  );
}

export function findHospitalByNumericId(numericId: number): Hospital | undefined {
  return getHospitals().find(h => h.numericId === numericId);
}

export function findDoctorsByHospitalNumericId(numericId: number): Doctor[] {
  const hospital = findHospitalByNumericId(numericId);
  return hospital?.doctors || [];
}

File: backend/src/main/java/com/icare/service/ExcelService.java
package com.icare.service;

import com.icare.model.Doctor;
import com.icare.model.Hospital;
import org.apache.poi.ss.usermodel.Cell;
import org.apache.poi.ss.usermodel.Row;
import org.apache.poi.ss.usermodel.Sheet;
import org.apache.poi.ss.usermodel.Workbook;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;
import org.springframework.core.io.ClassPathResource;
import org.springframework.stereotype.Service;

import jakarta.annotation.PostConstruct;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class ExcelService {

    private static final String HOSPITALS_RESOURCE = "data/Hospital_DB.xlsx";
    private static final String REFERENCE_SHEET_NAME = "Reference Table";

    private List<Hospital> hospitals = Collections.emptyList();
    private Map<String, List<Doctor>> doctorsByReferenceKey = Collections.emptyMap();

    @PostConstruct
    public void init() {
        this.hospitals = loadHospitals();
    }

    public List<Hospital> getHospitals() {
        return new ArrayList<>(hospitals);
    }

    public Optional<Hospital> findHospitalById(String hospitalId) {
        if (hospitalId == null) {
            return Optional.empty();
        }
        return hospitals.stream()
                .filter(hospital -> hospitalId.equals(hospital.getId()) ||
                        (hospital.getNumericId() != null && hospitalId.equals(String.valueOf(hospital.getNumericId()))))
                .findFirst();
    }

    public Optional<Hospital> findHospitalByNumericId(int numericId) {
        return hospitals.stream()
                .filter(hospital -> hospital.getNumericId() != null && hospital.getNumericId() == numericId)
                .findFirst();
    }

    public List<Doctor> findDoctorsByHospitalNumericId(int numericId) {
        return findHospitalByNumericId(numericId)
                .map(Hospital::getDoctors)
                .orElse(Collections.emptyList());
    }

    private List<Hospital> loadHospitals() {
        try (InputStream stream = new ClassPathResource(HOSPITALS_RESOURCE).getInputStream();
             Workbook workbook = new XSSFWorkbook(stream)) {

            Map<String, Sheet> sheets = buildSheetMap(workbook);
            Sheet referenceSheet = findReferenceSheet(sheets);
            if (referenceSheet == null) {
                return Collections.emptyList();
            }

            Map<String, List<Doctor>> doctorsByReference = loadDoctorSheets(sheets);
            this.doctorsByReferenceKey = doctorsByReference;

            List<Hospital> loaded = loadReferenceTable(referenceSheet, doctorsByReference);
            return loaded;
        } catch (IOException exception) {
            return Collections.emptyList();
        }
    }

    private Map<String, Sheet> buildSheetMap(Workbook workbook) {
        Map<String, Sheet> map = new HashMap<>();
        for (int index = 0; index < workbook.getNumberOfSheets(); index++) {
            Sheet sheet = workbook.getSheetAt(index);
            if (sheet != null) {
                map.put(sheet.getSheetName().trim(), sheet);
            }
        }
        return map;
    }

    private Sheet findReferenceSheet(Map<String, Sheet> sheets) {
        if (sheets.containsKey(REFERENCE_SHEET_NAME)) {
            return sheets.get(REFERENCE_SHEET_NAME);
        }
        for (Sheet sheet : sheets.values()) {
            Row headerRow = sheet.getRow(0);
            if (headerRow != null) {
                String header = normalizeHeader(safeString(headerRow.getCell(0)));
                if (header.contains("nameofhealthcarefacility") || header.contains("healthcarefacility")) {
                    return sheet;
                }
            }
        }
        return null;
    }

    private List<Hospital> loadReferenceTable(Sheet sheet, Map<String, List<Doctor>> doctorsByReference) {
        Row headerRow = sheet.getRow(0);
        List<String> headers = buildHeaderNames(headerRow);
        List<Hospital> loaded = new ArrayList<>();

        for (int rowIndex = 1; rowIndex <= sheet.getLastRowNum(); rowIndex++) {
            Row row = sheet.getRow(rowIndex);
            if (row == null) {
                continue;
            }
            Hospital hospital = parseReferenceRow(row, headers, rowIndex, doctorsByReference);
            if (hospital != null) {
                loaded.add(hospital);
            }
        }
        return loaded;
    }

    private Hospital parseReferenceRow(Row row, List<String> headers, int rowIndex, Map<String, List<Doctor>> doctorsByReference) {
        if (isReferenceRowEmpty(row)) {
            return null;
        }

        String referenceKey = safeString(row.getCell(getColumnIndex(headers, "referencekey", 5)));
        String name = safeString(row.getCell(getColumnIndex(headers, "nameofhealthcarefacility", 0)));
        if (name.isEmpty()) {
            name = safeString(row.getCell(getColumnIndex(headers, "name", 0)));
        }
        // Skip rows with no hospital name
        if (name.isEmpty()) {
            return null;
        }

        String type = safeString(row.getCell(getColumnIndex(headers, "type", 1)));
        String level = safeString(row.getCell(getColumnIndex(headers, "level", 2)));
        double latitude = safeDouble(row.getCell(getColumnIndex(headers, "latitude", 3)));
        double longitude = safeDouble(row.getCell(getColumnIndex(headers, "longitude", 4)));
        String address = safeString(row.getCell(getColumnIndex(headers, "address", 2)));

        // Skip rows with no valid coordinates — they cannot be placed on the map
        if (latitude == 0.0 && longitude == 0.0) {
            return null;
        }

        List<String> services = Collections.emptyList();

        Hospital hospital = new Hospital(referenceKey.isEmpty() ? String.valueOf(rowIndex) : referenceKey,
                name,
                address,
                latitude,
                longitude,
                services,
                "",
                "",
                "",
                referenceKey,
                level);
        hospital.setNumericId(rowIndex);

        List<Doctor> doctors = doctorsByReference.getOrDefault(referenceKey, Collections.emptyList());
        hospital.setDoctors(doctors);
        hospital.setServices(buildHospitalServices(doctors, type));

        return hospital;
    }

    private List<String> buildHospitalServices(List<Doctor> doctors, String hospitalType) {
        if (doctors == null || doctors.isEmpty()) {
            return Collections.singletonList(hospitalType != null && !hospitalType.isBlank() ? hospitalType : "General");
        }
        return doctors.stream()
                .map(Doctor::getDepartment)
                .filter(dep -> dep != null && !dep.isBlank())
                .map(String::trim)
                .distinct()
                .collect(Collectors.toList());
    }

    private boolean isReferenceRowEmpty(Row row) {
        if (row == null) {
            return true;
        }
        short lastCell = row.getLastCellNum();
        if (lastCell < 0) {
            return true;
        }
        for (int i = 0; i < lastCell; i++) {
            if (!safeString(row.getCell(i)).isBlank()) {
                return false;
            }
        }
        return true;
    }

    private Map<String, List<Doctor>> loadDoctorSheets(Map<String, Sheet> sheets) {
        Map<String, List<Doctor>> doctorsByReference = new HashMap<>();
        for (Map.Entry<String, Sheet> entry : sheets.entrySet()) {
            String sheetName = entry.getKey();
            Sheet sheet = entry.getValue();
            if (sheetName.equalsIgnoreCase(REFERENCE_SHEET_NAME)) {
                continue;
            }
            if (isDoctorSheet(sheet)) {
                List<Doctor> doctors = parseDoctorSheet(sheet);
                if (!doctors.isEmpty()) {
                    doctorsByReference.put(sheetName, doctors);
                }
            }
        }
        return doctorsByReference;
    }

    private boolean isDoctorSheet(Sheet sheet) {
        Row headerRow = sheet.getRow(0);
        if (headerRow == null) {
            return false;
        }
        String firstCell = normalizeHeader(safeString(headerRow.getCell(0)));
        return firstCell.contains("nameofdoctor") || firstCell.contains("doctor");
    }

    private List<Doctor> parseDoctorSheet(Sheet sheet) {
        Row headerRow = sheet.getRow(0);
        List<String> headers = buildHeaderNames(headerRow);
        List<Doctor> doctors = new ArrayList<>();

        for (int rowIndex = 1; rowIndex <= sheet.getLastRowNum(); rowIndex++) {
            Row row = sheet.getRow(rowIndex);
            if (row == null) {
                continue;
            }
            String name = safeString(row.getCell(getColumnIndex(headers, "nameofdoctor", 0)));
            if (name.isBlank()) {
                continue;
            }

            Doctor doctor = new Doctor();
            doctor.setId(generateDoctorId(sheet.getSheetName(), name, safeString(row.getCell(getColumnIndex(headers, "department", 2)))));
            doctor.setName(name);
            doctor.setType(safeString(row.getCell(getColumnIndex(headers, "typeofdoctor", 1))));
            doctor.setDepartment(safeString(row.getCell(getColumnIndex(headers, "department", 2))));
            String otherSpecializations = safeString(row.getCell(getColumnIndex(headers, "otherspecializations", 3)));
            doctor.setSchedule(safeString(row.getCell(getColumnIndex(headers, "schedule", 4))));
            doctor.setPriceRange(safeString(row.getCell(getColumnIndex(headers, "price", 7))));
            doctor.setHmos(parseHmos(row.getCell(getColumnIndex(headers, "hmos", 6))));
            doctor.setContact(safeString(row.getCell(getColumnIndex(headers, "contactnumber", 9))));
            doctor.setSecretary(safeString(row.getCell(getColumnIndex(headers, "medicalsecretary", 8))));
            doctor.setTags(buildDoctorTags(doctor.getDepartment(), otherSpecializations, doctor.getType()));

            doctor.setType(doctor.getType().isBlank() ? "General Practitioner" : doctor.getType());
            doctors.add(doctor);
        }

        return doctors;
    }

    private int generateDoctorId(String sheetName, String name, String department) {
        return Math.abs(Objects.hash(sheetName, name, department));
    }

    private List<String> buildDoctorTags(String department, String specializations, String type) {
        List<String> tags = new ArrayList<>();
        if (department != null && !department.isBlank()) {
            tags.addAll(Arrays.stream(department.split(","))
                    .map(String::trim)
                    .filter(s -> !s.isBlank())
                    .map(this::normalizeTag)
                    .collect(Collectors.toList()));
        }
        if (specializations != null && !specializations.isBlank()) {
            tags.addAll(Arrays.stream(specializations.split(","))
                    .map(String::trim)
                    .filter(s -> !s.isBlank())
                    .map(this::normalizeTag)
                    .collect(Collectors.toList()));
        }
        if (type != null && !type.isBlank()) {
            tags.add(normalizeTag(type));
        }
        if (tags.isEmpty()) {
            tags.add("general");
        }
        return tags.stream().distinct().collect(Collectors.toList());
    }

    private String normalizeTag(String value) {
        return value.trim().toLowerCase(Locale.ENGLISH).replaceAll("[^a-z0-9]+", "-");
    }

    private int getColumnIndex(List<String> headers, String fieldName, int fallbackIndex) {
        int index = headers.indexOf(fieldName);
        return index >= 0 ? index : fallbackIndex;
    }

    private List<String> buildHeaderNames(Row headerRow) {
        if (headerRow == null) {
            return Collections.emptyList();
        }
        List<String> headers = new ArrayList<>();
        for (int cellIndex = 0; cellIndex < headerRow.getLastCellNum(); cellIndex++) {
            headers.add(normalizeHeader(safeString(headerRow.getCell(cellIndex))));
        }
        return headers;
    }

    private String normalizeHeader(String raw) {
        if (raw == null) {
            return "";
        }
        return raw.toLowerCase(Locale.ENGLISH)
                .replace(" ", "")
                .replace("_", "")
                .replace("-", "");
    }

    private String safeString(Cell cell) {
        if (cell == null) {
            return "";
        }
        switch (cell.getCellType()) {
            case STRING:
                return cell.getStringCellValue().trim();
            case NUMERIC:
                return String.valueOf(cell.getNumericCellValue()).trim();
            case BOOLEAN:
                return String.valueOf(cell.getBooleanCellValue()).trim();
            default:
                return "";
        }
    }

    private double safeDouble(Cell cell) {
        if (cell == null) {
            return 0.0; // missing cell — caller will skip rows with 0.0 lat/lng
        }
        if (cell.getCellType() == org.apache.poi.ss.usermodel.CellType.NUMERIC) {
            return cell.getNumericCellValue();
        }
        String text = safeString(cell);
        if (text.isEmpty()) {
            return 0.0;
        }
        try {
            return Double.parseDouble(text);
        } catch (NumberFormatException e) {
            return 0.0; // unparseable — treat as missing
        }
    }

    private List<String> parseHmos(Cell hmosCell) {
        String raw = safeString(hmosCell);
        if (raw.isEmpty()) {
            return Collections.emptyList();
        }
        return Arrays.stream(raw.split(","))
                .map(String::trim)
                .filter(value -> !value.isEmpty())
                .collect(Collectors.toList());
    }
}
